macos/FluentUI/Button.swift

```diff
@@ -0,0 +1,230 @@
+		initialize()
+	}
+
+	@available(*, unavailable)
+	required public init?(coder decoder: NSCoder) {
+		preconditionFailure()
+	}
+	
+	private func initialize() {
+		// Do common initialization work
+		isBordered = false
+		wantsLayer = true
+		update()
+	}
+	
+	override public var image: NSImage? {
+		didSet {
+			if imagePosition != .imageOnly {
+				preconditionFailure("We currently do not support adding both an image and a title. Please set the imagePosition to .imageOnly before setting an image")
+			}
+		}
+	}
+	
+	override public func updateTrackingAreas() {
+		super.updateTrackingAreas()
+		
+		// Remove existing trackingArea
+		if let trackingArea = trackingArea {
+			removeTrackingArea(trackingArea)
+			self.trackingArea = nil
+		}
+		
+		// Create a new trackingArea
+		let opts: NSTrackingArea.Options = [.mouseEnteredAndExited, .activeAlways]
+		let trackingArea = NSTrackingArea(
+			rect: bounds,
+			options: opts,
+			owner: self,
+			userInfo: nil
+		)
+		addTrackingArea(trackingArea)
+		self.trackingArea = trackingArea
+	}
+	
+	override public func mouseEntered(with event: NSEvent) {
+		mouseEntered = true
+		needsDisplay = true
+	}
+
+	override public func mouseExited(with event: NSEvent) {
+		mouseEntered = false
+		needsDisplay = true
+	}
+	
+	override public func mouseDown(with event: NSEvent) {
+		mouseDown = true
+		needsDisplay = true
+	}
+	
+	override public func mouseUp(with event: NSEvent) {
+		mouseDown = false
+		needsDisplay = true
+	}
+	
+	open override func updateLayer() {
+		if let layer = layer {
+			layer.borderWidth = borderWidth
+			layer.masksToBounds = true
+			layer.cornerRadius = cornerRadius
+			layer.backgroundColor = layerBackgroundColor.cgColor
+			layer.borderColor = outlineColor.cgColor
+		}
+	}
+	
+	open override var intrinsicContentSize: NSSize {
+		var intrinsicContentSize = super.intrinsicContentSize;
+
+		intrinsicContentSize.width = intrinsicContentSize.width + (horizontalPadding * 2)
+		intrinsicContentSize.height += (verticalPadding * 2);
+
+		return intrinsicContentSize;
+	}
+	
+	override public var wantsUpdateLayer: Bool {
+		return true
+	}
+	
+	/// The primary color of the button, AKA, the fill color in the primaryFilled style, and the outline in the primaryOutline style
+	@objc public var primaryColor: NSColor = defaultPrimaryColor {
+		didSet {
+			if primaryColor != oldValue {
+				update()
+			}
+		}
+	}
+
+	/// The primary style of the button
+	@objc public var style: ButtonStyle = .primaryFilled {
+		didSet {
+			if style != oldValue {
+				update()
+			}
+		}
+	}
+
+	private func update() {
+		self.attributedTitle = NSAttributedString(string: title, attributes: [.foregroundColor : textColor])
+		needsDisplay = true
+	}
+	
+	private func disabledColor(color: NSColor) -> NSColor {
+		if #available(macOS 10.14, *) {
+			return color.withSystemEffect(.disabled)
+		} else {
+			return color.withAlphaComponent(0.25)
+		}
+	}
+	
+	private static var defaultPrimaryColor: NSColor {
```

Comment:
since this is static, might move it down to be more a constant and not in the middle of a bunch of instance configuration?